std-bits/base/src/main/java/org/enso/base/parser/Separators.java

```diff
@@ -0,0 +1,130 @@
+package org.enso.base.parser;
+
+import static org.enso.base.parser.NumberWithSeparators.isDigit;
+
+import java.nio.CharBuffer;
+
+public record Separators(char first, char second, int count, int endIdx, int lastSeparator) {
+  /** Check if the character is a separator. */
+  static boolean isSeparator(char c) {
+    return c == '.' || c == ',' || c == ' ' || c == '\'';
+  }
+
+  /**
+   * Strip out the specified separators and replace with just full stop for decimal. If any
+   * character other than a digit, thousands or decimal separator is encountered then return null.
+   * If multiple decimal separators are encountered then return null.
+   */
+  static CharSequence strip(CharSequence value, int idx, int endIdx, char thousands, char decimal) {
+    boolean foundDecimal = false;
+    char[] results = new char[endIdx - idx];
+    int resultIdx = 0;
+    for (int i = idx; i < endIdx; i++) {
+      char c = value.charAt(i);
+      if (c == decimal) {
+        if (foundDecimal) {
+          return null;
+        }
+        results[resultIdx++] = '.';
+        foundDecimal = true;
+      } else if (isDigit(c)) {
+        results[resultIdx++] = c;
+      } else if (c != thousands) {
+        return null;
+      }
+    }
+    return CharBuffer.wrap(results, 0, resultIdx);
+  }
+
+  private static boolean validChar(char c, char first, char second) {
+    return isDigit(c)
+        || (first == NumberWithSeparators.Constants.NONE
+            ? isSeparator(c)
+            : (second == NumberWithSeparators.Constants.NONE
+                ? c == first || c == ',' || c == '.'
+                : c == second));
+  }
+
+  /**
+   * Find the number and separators section. Validate the spacing of separators. Return the
+   * separators found or null if invalid.
+   */
+  static Separators parse(CharSequence value, int idx, boolean integer) {
+    int endIdx = idx;
+    char firstSeparator = NumberWithSeparators.Constants.NONE;
+    char secondSeparator = NumberWithSeparators.Constants.NONE;
+
+    boolean firstWasSeparator = false;
+    int lastSeparator = -1;
+    int separatorCount = 0;
+
+    // Scan the text, find and validate spacing of separators.
+    // Space and ' are both valid thousands separators, but can't be second separator.
+    // ToDo: Cope with scientific notation.
+    char c = value.charAt(idx);
+    while (endIdx < value.length() && validChar(c, firstSeparator, secondSeparator)) {
```

Comment:
this is a stylistic preference, but it feels to me like a for loop could be clearer here, or at the very least moving the `validChar` into the inside of the loop


I'd probably do it as
```java
for (char c = value.charAt(endIdx); endIdx < value.length(); endIdx++) {
    if (!validChar(c, firstSeparator, secondSeparator)) {
       break;
    }
}
```

As it now makes it clear what we are iterating over in the loop and what the conditions are. What I don't like about the current approach is that the conditions are spread apart - part of them is at the top in `while(COND)` but the update condition is at the end of the inner loop body - which is far away because the loop body is large. So I cannot easily glance what this loop is iterating over, because the inner workings on the loop obscure that.